opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -130,6 +134,68 @@ where
             );
         }
     }
+
+    /// Iterate through all attribute sets and populate `DataPoints` in readonly mode.
+    /// This is used in Cumulative temporality mode, where [`ValueMap`] is not cleared.
+    pub(crate) fn collect_readonly<Res, MapFn>(&self, dest: &mut Vec<Res>, mut map_fn: MapFn)
+    where
+        MapFn: FnMut(Vec<KeyValue>, &A) -> Res,
+    {
+        prepare_data(dest, self.count.load(Ordering::SeqCst));
+        if self.has_no_attribute_value.load(Ordering::Acquire) {
+            dest.push(map_fn(vec![], &self.no_attribute_tracker));
+        }
+
+        let Ok(trackers) = self.trackers.read() else {
+            return;
+        };
+
+        let mut seen = HashSet::new();
+        for (attrs, tracker) in trackers.iter() {
+            if seen.insert(Arc::as_ptr(tracker)) {
+                dest.push(map_fn(attrs.clone(), tracker));
+            }
+        }
+    }
+
+    /// Iterate through all attribute sets, populate `DataPoints` and reset.
+    /// This is used in Delta temporality mode, where [`ValueMap`] is reset after collection.
+    pub(crate) fn collect_and_reset<Res, MapFn>(&self, dest: &mut Vec<Res>, mut map_fn: MapFn)
+    where
+        MapFn: FnMut(Vec<KeyValue>, A) -> Res,
+    {
+        prepare_data(dest, self.count.load(Ordering::SeqCst));
+        if self.has_no_attribute_value.swap(false, Ordering::AcqRel) {
+            dest.push(map_fn(
+                vec![],
+                self.no_attribute_tracker.clone_and_reset(&self.config),
+            ));
+        }
+
+        let trackers = match self.trackers.write() {
+            Ok(mut trackers) => {
+                self.count.store(0, Ordering::SeqCst);
+                take(trackers.deref_mut())
```

Comment:
This would lead to additional allocation as we would be replacing `trackers` with a new `Hashmap` (with capacity - 0) on each collect. So, after every collect, we would be allocating to add entries to the Hashmap. This was avoided earlier when we used `drain` on `Hashmap`, which retained the Hashmap's capacity.

One approach to solve this would be to add another `RwLock<HashMap>` field to the `ValueMap` struct and instead of calling `mem::take` here, we could `mem::swap` with current trackers with this newly added Hashmap.

Reply:
I tried to minimize holding write lock, as much as possible (i cannot release lock if I want to use 'drain').
In #2247 I have observed almost 4x improvement (when collect phase is running), with this change.
Regarding extra allocation, I don't think that it makes a huge difference, (given that they happen constantly, and we really don't notice them that much:)). 
Maybe I could improve this just a bit by doing something like `replace(trackers, Hashmap::with_capacity(n))`? and leave other potential improvements for the future.
Would that work for you?

Reply:
I'm totally in favor of reducing the time held for write lock, so I'm not asking to go back to calling `drain` with the write lock. However, since we are optimizing collect anyway why not just avoid any amount of allocation that we can?

One simple way to fix this would be:

1. Add another field `trackers_for_collect` to the `ValueMap` struct

```rust
pub(crate) struct ValueMap<A>
where
    A: Aggregator,
{
    /// Trackers store the values associated with different attribute sets.
    trackers: RwLock<HashMap<Vec<KeyValue>, Arc<A>>>,

    /// Trackers store the values associated with different attribute sets.
    trackers_for_collect: RwLock<HashMap<Vec<KeyValue>, Arc<A>>>,

    ...
}

impl<A> ValueMap<A>
where
    A: Aggregator,
{
    fn new(config: A::InitConfig) -> Self {
        ValueMap {
            trackers: RwLock::new(HashMap::new()),
            trackers_for_collect: RwLock::new(HashMap::new()),

            ...
        }
    }
}
```

2. Update the `collect_and_reset` method to:
   - Use `mem::swap` to swap the trackers used for updates with the trackers used for collect.
   - `trackers_for_collect` would then have the aggregated data
   - Call `drain` instead of `into_iter` on trackers for collect to retain its capacity for subsequent collects

```rust
pub(crate) fn collect_and_reset<Res, MapFn>(&self, dest: &mut Vec<Res>, mut map_fn: MapFn)
    where
        MapFn: FnMut(Vec<KeyValue>, A) -> Res,
{
    prepare_data(dest, self.count.load(Ordering::SeqCst));

    ...
    
    let mut trackers = match self.trackers.write() {
        Ok(mut trackers_guard) => {
            self.count.store(0, Ordering::SeqCst);
            // take(trackers_guard.deref_mut())
            let mut trackers_export = self.trackers_for_collect.write().unwrap();
            swap(trackers_guard.deref_mut(), trackers_export.deref_mut()); // swap the tracker instances
            trackers_export
        }
        Err(_) => todo!(),
    };

    let mut seen = HashSet::new();
    for (attrs, tracker) in trackers.drain() { // calling `drain` here retains the capacity of `trackers_for_collect`
        if seen.insert(Arc::as_ptr(&tracker)) {
            ...
        }
    }
}
```

How does that look?

Reply:
nit: if we go with the 2 hashmap approach, leave a code comment like "even though trackers_for_collect is RWLock guarded, only collect() thread ever need to accesses it, so the W lock on it has no impact on update() threads.".

Reply:
Agreed to come back to this in next PR.